internal/controllers/__snapshots__/workspace_controller_test.snap

```diff
@@ -245,30 +245,7 @@ SnapShot = """
     ]
   },
   \"spec\": {
-    \"routes\": [
-      {
-        \"match\": \"PathPrefix(`/path`)\",
-        \"kind\": \"Rule\",
-        \"priority\": 100,
-        \"services\": [
-          {
-            \"name\": \"ws-test-ws-svc\",
-            \"kind\": \"Service\",
-            \"port\": 3000,
-            \"scheme\": \"http\"
-          }
-        ],
-        \"middlewares\": [
-          {
-            \"name\": \"userNameHeader\"
-          },
-          {
-            \"name\": \"cosmo-auth\",
-            \"namespace\": \"cosmo-system\"
-          }
-        ]
-      }
-    ],
+    \"routes\": [],
```

Comment:
Empty routes are correct in this test case?

Reply:
I get this result when I run the test after #745 merge.
Is this unexpected?

Reply:
Empty seems to be wrong.
Please run the test there as well.

Reply:
The snapshot is correct. I did but no change.
Do you run it at the current main branch?
Please check your test logs and your environment?

Reply:
I tested it in a new workspace and it matched.
When I git clone the latest code in the existing workspace, it does not match. Attached is the log.
```
------------------------------
• [FAILED] [0.063 seconds]
Workspace controller when updating Workspace spec [It] should do reconcile again and update child Instance
/projects/cosmo_cur2/internal/controllers/workspace_controller_test.go:189

  Timeline >>
  2023-06-08T09:01:20+09:00     ERROR   Reconciler error        {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "e851be99-6774-459e-a2f7-8380e3ff6f4d", "error": "Operation cannot be fulfilled on workspaces.cosmo-workspace.github.io \"ws-test\": the object has been modified; please apply your changes to the latest version and try again"}
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /home/coder/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:329
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /home/coder/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /home/coder/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceReconciler     start reconcile {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "f4614f92-d946-4038-bc6c-ba3b642b5399", "req": "cosmo-user-wsctltest/ws-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceReconciler     Object diff:   &v1alpha1.Workspace{
        TypeMeta:   {Kind: "Workspace", APIVersion: "cosmo-workspace.github.io/v1alpha1"},
        ObjectMeta: {Name: "ws-test", Namespace: "cosmo-user-wsctltest", UID: "752fef54-50cf-4963-8001-bc2555cd58d3", ResourceVersion: "298", ...},
        Spec: v1alpha1.WorkspaceSpec{
                Template: {Name: "code-server-test"},
                Replicas: &1,
                Vars: map[string]string{
                        "{{DOMAIN}}":                           "example.com",
                        "{{IMAGE_TAG}}":                        "latest",
  +                     "{{USER_NAME}}":                        "wsctltest",
  +                     "{{WORKSPACE_DEPLOYMENT_NAME}}":        "ws-dep",
  +                     "{{WORKSPACE_INGRESS_NAME}}":           "ws-ing",
  +                     "{{WORKSPACE_SERVICE_MAIN_PORT_NAME}}": "mainPort",
  +                     "{{WORKSPACE_SERVICE_NAME}}":           "ws-svc",
  +                     "{{WORKSPACE}}":                        "ws-test",
                },
                Network: nil,
        },
        Status: v1alpha1.WorkspaceStatus{
                Instance: v1alpha1.ObjectRef{
                        ObjectReference: v1.ObjectReference{
  -                             Kind:            "",
  +                             Kind:            "Instance",
  -                             Namespace:       "",
  +                             Namespace:       "cosmo-user-wsctltest",
  -                             Name:            "",
  +                             Name:            "ws-test",
  -                             UID:             "",
  +                             UID:             "644df7c0-da8b-417b-9d42-75b3246a6c07",
  -                             APIVersion:      "",
  +                             APIVersion:      "cosmo-workspace.github.io/v1alpha1",
  -                             ResourceVersion: "",
  +                             ResourceVersion: "309",
                                FieldPath:       "",
                        },
  -                     CreationTimestamp: nil,
  +                     CreationTimestamp: s"2023-06-08 09:01:20 +0900 JST",
                },
                Phase:  "Starting",
                URLs:   nil,
                Config: {DeploymentName: "ws-dep", ServiceName: "ws-svc", IngressName: "ws-ing", ServiceMainPortName: "mainPort", ...},
        },
    }
        {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "f4614f92-d946-4038-bc6c-ba3b642b5399", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test", "x": "ws-test", "y": "ws-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceStatusReconciler       start reconcile {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "77ee6247-8319-4f6d-a18f-bdbd020212ab", "req": "cosmo-user-wsctltest/ws-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceStatusReconciler       workspace urlmap: map[] {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "77ee6247-8319-4f6d-a18f-bdbd020212ab", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceStatusReconciler       finish reconcile        {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "77ee6247-8319-4f6d-a18f-bdbd020212ab", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test", "requeue": true}
  2023-06-08T09:01:20+09:00     INFO    WorkspaceReconciler     status updated  {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "f4614f92-d946-4038-bc6c-ba3b642b5399", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceReconciler     finish reconcile        {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "f4614f92-d946-4038-bc6c-ba3b642b5399", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceReconciler     start reconcile {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "5ddfd4a7-7ef8-4e6a-a47c-5870fde9c690", "req": "cosmo-user-wsctltest/ws-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceReconciler     Object diff:   &v1alpha1.Workspace{
        TypeMeta:   {Kind: "Workspace", APIVersion: "cosmo-workspace.github.io/v1alpha1"},
        ObjectMeta: {Name: "ws-test", Namespace: "cosmo-user-wsctltest", UID: "752fef54-50cf-4963-8001-bc2555cd58d3", ResourceVersion: "314", ...},
        Spec: v1alpha1.WorkspaceSpec{
                Template: {Name: "code-server-test"},
                Replicas: &1,
                Vars: map[string]string{
                        "{{DOMAIN}}":                           "example.com",
                        "{{IMAGE_TAG}}":                        "latest",
  +                     "{{USER_NAME}}":                        "wsctltest",
  +                     "{{WORKSPACE_DEPLOYMENT_NAME}}":        "ws-dep",
  +                     "{{WORKSPACE_INGRESS_NAME}}":           "ws-ing",
  +                     "{{WORKSPACE_SERVICE_MAIN_PORT_NAME}}": "mainPort",
  +                     "{{WORKSPACE_SERVICE_NAME}}":           "ws-svc",
  +                     "{{WORKSPACE}}":                        "ws-test",
                },
                Network: nil,
        },
        Status: {Instance: {ObjectReference: {Kind: "Instance", Namespace: "cosmo-user-wsctltest", Name: "ws-test", UID: "644df7c0-da8b-417b-9d42-75b3246a6c07", ...}, CreationTimestamp: s"2023-06-08 09:01:20 +0900 JST"}, Phase: "Starting", Config: {DeploymentName: "ws-dep", ServiceName: "ws-svc", IngressName: "ws-ing", ServiceMainPortName: "mainPort", ...}},
    }
        {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "5ddfd4a7-7ef8-4e6a-a47c-5870fde9c690", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test", "x": "ws-test", "y": "ws-test"}
  2023-06-08T09:01:20+09:00     INFO    WorkspaceReconciler     status updated  {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "5ddfd4a7-7ef8-4e6a-a47c-5870fde9c690", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceReconciler     finish reconcile        {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "5ddfd4a7-7ef8-4e6a-a47c-5870fde9c690", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceReconciler     start reconcile {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "3a5043fc-d211-45eb-8815-3e43d54c05cf", "req": "cosmo-user-wsctltest/ws-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceStatusReconciler       start reconcile {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "733d90d1-94a5-4abb-97c1-a76b850a47a3", "req": "cosmo-user-wsctltest/ws-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceStatusReconciler       workspace urlmap: map[port1:https://group1-{{WOKRSPACE}}-{{USER}}.domain/path]  {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "733d90d1-94a5-4abb-97c1-a76b850a47a3", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test"}
  2023-06-08T09:01:20+09:00     DEBUG   WorkspaceStatusReconciler       Object diff:   &v1alpha1.Workspace{
        TypeMeta:   {Kind: "Workspace", APIVersion: "cosmo-workspace.github.io/v1alpha1"},
        ObjectMeta: {Name: "ws-test", Namespace: "cosmo-user-wsctltest", UID: "752fef54-50cf-4963-8001-bc2555cd58d3", ResourceVersion: "315", ...},
        Spec:       {Template: {Name: "code-server-test"}, Replicas: &0, Vars: {"{{DOMAIN}}": "example.com", "{{IMAGE_TAG}}": "latest"}, Network: {{Name: "port1", PortNumber: 3000, HTTPPath: "/path", TargetPortNumber: &30000, ...}}},
        Status: v1alpha1.WorkspaceStatus{
                Instance: {ObjectReference: {Kind: "Instance", Namespace: "cosmo-user-wsctltest", Name: "ws-test", UID: "644df7c0-da8b-417b-9d42-75b3246a6c07", ...}, CreationTimestamp: s"2023-06-08 09:01:20 +0900 JST"},
  -             Phase:    "Starting",
  +             Phase:    "Stopped",
  -             URLs:     nil,
  +             URLs:     map[string]string{"port1": "https://group1-{{WOKRSPACE}}-{{USER}}.domain/path"},
                Config:   {DeploymentName: "ws-dep", ServiceName: "ws-svc", IngressName: "ws-ing", ServiceMainPortName: "mainPort", ...},
        },
    }
        {"controller": "workspace", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Workspace", "Workspace": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "733d90d1-94a5-4abb-97c1-a76b850a47a3", "req": "cosmo-user-wsctltest/ws-test", "UID": "752fef54-50cf-4963-8001-bc2555cd58d3", "Template": "code-server-test", "x": "ws-test", "y": "ws-test"}
  2023-06-08T09:01:20+09:00     DEBUG   InstanceReconciler      start reconcile {"controller": "instance", "controllerGroup": "cosmo-workspace.github.io", "controllerKind": "Instance", "Instance": {"name":"ws-test","namespace":"cosmo-user-wsctltest"}, "namespace": "cosmo-user-wsctltest", "name": "ws-test", "reconcileID": "2276da69-f228-486d-96aa-5c0ae9d0956a", "req": "cosmo-user-wsctltest/ws-test"}
  2023-06-08T09:01:20+09:00     DEBUG   events  successfully reconciled. instance synced        {"type": "Normal", "object": {"kind":"Workspace","namespace":"cosmo-user-wsctltest","name":"ws-test","uid":"752fef54-50cf-4963-8001-bc2555cd58d3","apiVersion":"cosmo-workspace.github.io/v1alpha1","resourceVersion":"315"}, "reason": "updated"}
  [FAILED] in [It] - /projects/cosmo_cur2/internal/controllers/workspace_controller_test.go:228 @ 06/08/23 09:01:20.97
  << Timeline

  [FAILED] Expected to match
    (
        """
        ... // 18 identical lines
          },
          "spec": {
  -         "routes": [
  -           {
  -             "match": "PathPrefix(`/path`)",
  -             "kind": "Rule",
  -             "priority": 100,
  -             "services": [
  -               {
  -                 "name": "ws-test-ws-svc",
  -                 "kind": "Service",
  -                 "port": 3000,
  -                 "scheme": "http"
  -               }
  -             ],
  -             "middlewares": [
  -               {
  -                 "name": "userNameHeader"
  -               },
  -               {
  -                 "name": "cosmo-auth",
  -                 "namespace": "cosmo-system"
  -               }
  -             ]
  -           }
  -         ],
  +         "routes": [],
            "entryPoints": [
              "web",
        ... // 5 identical lines
        """
    )

  
  In [It] at: /projects/cosmo_cur2/internal/controllers/workspace_controller_test.go:228 @ 06/08/23 09:01:20.97
------------------------------
•••••••••••E0608 09:01:27.389064   15333 request.go:1075] Unexpected error when reading response body: context canceled
E0608 09:01:28.399900   15333 event.go:276] Unable to write event: '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"test-pv-clusterinst1.17668609db5660af", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:"ClusterInstance", Namespace:"", Name:"test-pv-clusterinst1", UID:"45144a8c-8ecd-4dee-821a-ee5e8e09a984", APIVersion:"cosmo-workspace.github.io/v1alpha1", ResourceVersion:"388", FieldPath:""}, Reason:"SyncFailed", Message:"failed to get resource: kind = StorageClass name = test-pv-clusterinst1-sc: client rate limiter Wait returned an error: context canceled", Source:v1.EventSource{Component:"cosmo-instance-controller", Host:""}, FirstTimestamp:time.Date(2023, time.June, 8, 9, 1, 27, 389528239, time.Local), LastTimestamp:time.Date(2023, time.June, 8, 9, 1, 27, 389528239, time.Local), Count:1, Type:"Warning", EventTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'Post "https://127.0.0.1:37369/api/v1/namespaces/default/events": dial tcp 127.0.0.1:37369: connect: connection refused'(may retry after sleeping)


Summarizing 1 Failure:
  [FAIL] Workspace controller when updating Workspace spec [It] should do reconcile again and update child Instance
  /projects/cosmo_cur2/internal/controllers/workspace_controller_test.go:228
```